src/odemis/acq/stream_settings.py

```diff
@@ -0,0 +1,174 @@
+
+from odemis import model, util
+from odemis.acq.stream import FluoStream
+from odemis.util.img import md_format_to_tint, tint_to_md_format
+
+SETTINGS_ORDER = ["name", "excitation", "power", "emission", "tint", "detExposureTime", "auto_bc_outliers"]
+NON_SETTINGS_VA = ['acquisitionType', 'auto_bc', 'background', 'histogram', 'image', 'intensityRange', 'is_active',
+                   'roi', 'should_update', 'single_frame_acquisition', 'status']
+
+
+def get_config_index(config_data: list, name: str) -> int:
+    """
+    Get the index of the config entry with the given name
+    :param config_data: list of config entries
+    :param name: name of config entry
+    :return: index of config entry with given name
+    """
+    try:
+        index = list(map(lambda x: x["name"], config_data)).index(name)
+    except:
+        index = None
+
+    return index
+
+
+class StreamSettings:
+    # TODO change the description
+    """
+    Read and write a list of most recently used (mru) settings from and to a JSON file respectively. The first element
+    in the list will be the latest used setting and the last element will be the oldest used setting. The list will
+    consist of finite number of entries which will update in the above mention order each time the list of settings
+    is updated.
+    """
+
+    def __init__(self, file_path, max_entries, config_data: List[Dict[str, Any]] = None):
+        """
+        :param file_path: full directory path to save the JSON file
+        :param max_entries: number of entries to save in the JSON file
+        :param config_data: list of stream controller settings where the first element is the latest
+         stream entry. The number of entries is defined by max_entries.
+        """
+        self.max_entries = max_entries
+        self.config_data = config_data
+        self.file_path = file_path
+        self._read()  # update the config_data by reading the JSON file
+        self.settings_order = None  # the ordered list of setting parameters used for loading any given stream
+
+    def _read(self):
+        logging.debug("Reading the most recently used stream settings in JSON")
+        if os.path.exists(self.file_path):
+            try:
+                with open(self.file_path, 'r') as jsonfile:
+                    self.config_data = json.load(jsonfile)
+            except json.decoder.JSONDecodeError as e:
+                logging.warning(f"Error decoding JSON: {e}")
+            except Exception as e:
+                logging.warning(f"Error in reading the stream settings: {e}")
+
+    def _write(self):
+        logging.debug("Writing the most recently used stream settings in JSON")
+        try:
+            if self.config_data:
+                with open(self.file_path, 'w') as jsonfile:
+                    json.dump(self.config_data, jsonfile, indent=2)
+        except Exception as e:
+            logging.warning(f"Error in writing the stream settings: {e}")
```

Comment:
The name of this function is `_write()`. If `write()` fails to write, it's a big deal typically! So I would rather have it pass the exception. The *caller* can decide that failure to write is not a big deal, and just log it, but not the function itself. => move the try/except to the caller. 

Also, I am pretty sure that there is no good reason it didn't work, so please just make it at `log.exception()`.

Same thing with `read()`. Though the failure to read at init could be just because there is no file existing yet, so it would just be log.info if the file doesn't exist.

Reply:
In def _read(self) functions, it checks for file path before reading the file. This function is called at init of the class, do I log the exception like below for read function ?
```suggestion
        try:
            self._read()  # update the config_data by reading the JSON file
        except Exception as e:
            logging.exception(f"Error in reading the stream settings: {e}")
```

Reply:
I think it should be considered normal that the file doesn't exist (as this happens on the first run). So I'd recommand to catch such error separately and just indicate at "info" level. For the other exceptions, "exception" is fine. You just don't need to include the content in the message, as that's what "exception()" does already.